crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
+// These are analagous to the ec_key functions in crypto/fipsmodule/ec/ec_key.c
 
-  if (pkey == NULL || ctx == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
-    goto err;
+typedef struct {
+  const PQDSA *pqdsa;
+} PQDSA_PKEY_CTX;
+
+PQDSA_KEY *PQDSA_KEY_new(void) {
+  PQDSA_KEY *ret = OPENSSL_zalloc(sizeof(PQDSA_KEY));
+  if (ret == NULL) {
+    return NULL;
   }
 
-  evp_pkey_set_method(pkey, &dilithium3_asn1_meth);
+  return ret;
+}
 
-  if (ml_dsa_65_keypair(key->pub, key->priv) != 0) {
-    goto err;
+static void PQDSA_KEY_clear(PQDSA_KEY *key) {
+  key->pqdsa = NULL;
+  OPENSSL_free(key->public_key);
+  OPENSSL_free(key->secret_key);
+  key->public_key = NULL;
+  key->secret_key = NULL;
+}
+
+int PQDSA_KEY_init(PQDSA_KEY *key, const PQDSA *pqdsa) {
+  if (key == NULL || pqdsa == NULL) {
+    return 0;
   }
+  // If the key is already initialized clear it.
+  PQDSA_KEY_clear(key);
 
-  OPENSSL_free(pkey->pkey.ptr);
-  pkey->pkey.ptr = key;
+  key->pqdsa = pqdsa;
+  key->public_key = OPENSSL_malloc(pqdsa->public_key_len);
+  key->secret_key = OPENSSL_malloc(pqdsa->secret_key_len);
+  if (key->public_key == NULL || key->secret_key == NULL) {
+    PQDSA_KEY_clear(key);
+    return 0;
+  }
   return 1;
+}
 
-err:
-  if (key != NULL) {
-    OPENSSL_free(key->pub);
-    OPENSSL_free(key->priv);
+void PQDSA_KEY_free(PQDSA_KEY *key) {
+  if (key == NULL) {
+    return;
   }
+  PQDSA_KEY_clear(key);
   OPENSSL_free(key);
-  return 0;
+}
 
+const PQDSA *PQDSA_KEY_get0_sig(PQDSA_KEY* key) {
+  return key->pqdsa;
 }
 
-static int pkey_dilithium3_sign_message(EVP_PKEY_CTX *ctx, uint8_t *sig,
-                                        size_t *siglen, const uint8_t *tbs,
-                                        size_t tbslen) {
-  if (ctx == NULL || ctx->pkey->pkey.ptr == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
+// PQDSA PKEY functions
+
+static int pkey_pqdsa_init(EVP_PKEY_CTX *ctx) {
+  PQDSA_PKEY_CTX *dctx;
+  dctx = OPENSSL_zalloc(sizeof(PQDSA_PKEY_CTX));
+  if (dctx == NULL) {
     return 0;
   }
 
-  DILITHIUM3_KEY *key = ctx->pkey->pkey.ptr;
+  ctx->data = dctx;
 
-  if (!key->priv) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+  return 1;
+}
+
+static void pkey_pqdsa_cleanup(EVP_PKEY_CTX *ctx) {
+  OPENSSL_free(ctx->data);
+}
+
+static int pkey_pqdsa_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  GUARD_PTR(ctx);
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  GUARD_PTR(dctx);
+  const PQDSA *pqdsa = dctx->pqdsa;
+  if (pqdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    pqdsa = PQDSA_KEY_get0_sig(ctx->pkey->pkey.pqdsa_key);
+  }
+
+  PQDSA_KEY *key = PQDSA_KEY_new();
+  if (key == NULL ||
+      !PQDSA_KEY_init(key, pqdsa) ||
+      !pqdsa->method->keygen(key->public_key, key->secret_key) ||
+      !EVP_PKEY_set_type(pkey, EVP_PKEY_NISTDSA)) {
+    PQDSA_KEY_free(key);
     return 0;
+      }
```

Comment:
```suggestion
  }
```

Reply:
fixed in [e2234e3](https://github.com/aws/aws-lc/pull/1963/commits/e2234e347025d64cd9f912fb78555585e03a6198)